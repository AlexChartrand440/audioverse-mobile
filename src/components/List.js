import React from 'react'
import PropTypes from 'prop-types'
import { FlatList, ActivityIndicator } from 'react-native'

const List = ({ renderItem, items, isFetching, onRefresh, onEndReached, selectItem, nextPageUrl }) => {

  if (!items.length) {
    return (
      <ActivityIndicator
        size="large"
        color="#03A9F4"
        style={{margin: 50}}
      />
    )
  }

  const handleEndReached = () => {
    console.log('end reached!!')
    !onEndReachedCalledDuringMomentum && nextPageUrl && typeof onEndReached === 'function' && onEndReached()
  }

  // issue with FlatList, onEndReached is triggered when there is few items
  // https://github.com/facebook/react-native/issues/14015
  let onEndReachedCalledDuringMomentum = true

  return (
    <FlatList
      data={items}
      renderItem={renderItem}
      keyExtractor={item => item.id}
      refreshing={isFetching}
      onRefresh={onRefresh}
      onEndReachedThreshold={0.1}
      onEndReached={handleEndReached}
      onMomentumScrollBegin={() => { onEndReachedCalledDuringMomentum = false }}
    />
  )

}

List.propTypes = {
  renderItem: PropTypes.func.isRequired,
  items: PropTypes.array.isRequired,
  isFetching: PropTypes.bool.isRequired,
  onEndReached: PropTypes.func,
  onRefresh: PropTypes.func.isRequired,
  nextPageUrl: PropTypes.string
}

export default List
