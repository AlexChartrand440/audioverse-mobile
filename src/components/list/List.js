import React from 'react'
import PropTypes from 'prop-types'
import {
  View,
  Text,
  Image,
  FlatList,
  ActivityIndicator,
  StyleSheet,
} from 'react-native'
import { ListItem } from 'react-native-elements'

import { defaultImage } from 'src/styles'

class List extends React.PureComponent {

  onEndReachedCalledDuringMomentum = true

  componentDidMount() {
    const { url, id } = this.props.navigation.state.params || {}
    this.props.actions.loadData(false, false, url, id)
  }

  handleEndReached = () => {
    console.log('end reached!!', this.onEndReachedCalledDuringMomentum)
    const { pagination } = this.props
    if (!this.onEndReachedCalledDuringMomentum && pagination && pagination.nextPageUrl) {
      this.onEndReachedCalledDuringMomentum = true
      const { url, id } = this.props.navigation.state.params || {}
      this.props.actions.loadData(true, false, url, id)
    }
  }

  handleRefresh = () => {
    const { url, id } = this.props.navigation.state.params || {}
    this.props.actions.loadData(false, true, url, id)
  }

  avatar(item) {
    const avatar = this.props.avatarExtractor(item)
    return {
      source: avatar && avatar.toString().startsWith('http') ? 
      { uri: avatar } : avatar ? avatar : defaultImage
    }
  }

  handlePress(item) {
    if (this.props.onPress) {
      this.props.onPress(item)
    } else {
      if (this.props.playlist) {
        this.props.localActions.resetAndPlayTrack(this.props.items, item.id)
      } else {
        this.props.localActions.resetAndPlayTrack([item])
      }
    }
  }

  renderItem = ({ item }) => {
    return (
      <ListItem
        leftAvatar={this.avatar(item)}
        title={this.props.titleExtractor(item)}
        titleProps={{numberOfLines: 1}}
        subtitle={this.props.subtitleExtractor(item)}
        onPress={() => this.handlePress(item)}
        bottomDivider
      />
    )
  }
  
  render() {
    const { navigation, items, pagination, renderItem, keyExtractor } = this.props

    if (!items.length && pagination.isFetching) {
      return (
        <ActivityIndicator
          size="large"
          color="#03A9F4"
          style={{margin: 50}}
        />
      )
    }

    const { image, description } = navigation.state.params || {}
    const Header =
      image && description ?
        <View style={styles.headerContainer}>
          <View style={styles.headerImageContainer}>
            <Image
              style={styles.headerImage}
              source={{uri: image}}
            />
          </View>
          { description ? <Text style={styles.headerText}>{description}</Text> : null }
        </View>
        : null
      

    return (
      <View style={styles.container}>
        <FlatList
          ListHeaderComponent={Header}
          data={items}
          renderItem={renderItem ? renderItem : this.renderItem}
          keyExtractor={keyExtractor}
          refreshing={pagination.isFetching}
          onRefresh={this.handleRefresh}
          onEndReachedThreshold={0.1}
          onEndReached={this.handleEndReached}
          onMomentumScrollBegin={() => { this.onEndReachedCalledDuringMomentum = false }}
        />
      </View>
    )
  }

}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  headerContainer: {
    padding: 20,
    backgroundColor: '#CCCCCC50',
    textAlign: 'center'
  },
  headerImageContainer: {
    alignItems: 'center'
  },
  headerImage: {
    width: 128,
    height: 128,
    borderRadius: 64
  },
  headerText: {
    textAlign: 'justify',
    marginTop: 10,
  },
})

List.propTypes = {
  navigation: PropTypes.object.isRequired,
  items: PropTypes.array,
  pagination: PropTypes.object,
  renderItem: PropTypes.func,
  keyExtractor: PropTypes.func,
  avatarExtractor: PropTypes.func,
  titleExtractor: PropTypes.func,
  subtitleExtractor: PropTypes.func,
  onPress: PropTypes.func,
  localActions: PropTypes.shape({
    resetAndPlayTrack: PropTypes.func.isRequired,
  }),
  actions: PropTypes.shape({
    loadData: PropTypes.func.isRequired,
  }),
}

List.defaultProps = {
  keyExtractor: item => item.id,
  avatarExtractor: item => item.artwork,
  titleExtractor: item => item.title,
  subtitleExtractor: item => item.artist + ' \u00B7 ' + item.duration,
  playlist: false,
}

export default List
