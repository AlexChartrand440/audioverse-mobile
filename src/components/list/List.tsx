import React, { useEffect, useRef } from 'react'
import {
  View,
  Text,
  Image,
  FlatList,
  ActivityIndicator,
  StyleSheet,
  ListRenderItem,
} from 'react-native'
import { NavigationInjectedProps } from 'react-navigation'
import { ListItem } from 'react-native-elements'
import { Track } from 'react-native-track-player'

import { defaultImage } from '../../styles'
import { PaginationState } from '../../store/paginate'

interface Item {
  [key: string]: any
}

interface Props extends NavigationInjectedProps {
  items: any[]
  pagination: PaginationState
  renderItem?: ListRenderItem<Item>
  keyExtractor?: (item: Item) => string
  avatarExtractor?: (item: Item) => string
  titleExtractor?: (item: Item) => string
  subtitleExtractor?: (item: Item) => string
  playlist?: boolean
  onPress?: (item: Item) => void
  localActions: {
    resetAndPlayTrack: (tracks: Track[], id: string | null) => void
  }
  actions: {
    loadData: (loadMore: boolean, refresh: boolean, url: string, id: string) => void
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  headerContainer: {
    padding: 20,
    backgroundColor: '#CCCCCC50',
    textAlign: 'center'
  },
  headerImageContainer: {
    alignItems: 'center'
  },
  headerImage: {
    width: 128,
    height: 128,
    borderRadius: 64
  },
  headerText: {
    textAlign: 'justify',
    marginTop: 10,
  },
})

export const List: React.FC<Props> = ({
  navigation,
  items,
  pagination,
  renderItem,
  keyExtractor = (item) => item.id,
  avatarExtractor = (item) => item.artwork,
  titleExtractor = (item) => item.title,
  subtitleExtractor = (item) => item.artist + ' \u00B7 ' + item.durationFormatted,
  onPress,
  playlist = false,
  localActions,
  actions,
}) => {

  const onEndReachedCalledDuringMomentumRef = useRef(true)

  useEffect(() => {
    const { url, id } = navigation.state.params || { url: null, id: null }
    actions.loadData(false, false, url, id)
  }, [])

  if (!items.length && pagination.isFetching) {
    return (
      <ActivityIndicator
        size="large"
        color="#03A9F4"
        style={{margin: 50}}
      />
    )
  }

  const handleEndReached = () => {
    console.log('end reached!!', onEndReachedCalledDuringMomentumRef.current)
    if (!onEndReachedCalledDuringMomentumRef.current && pagination && pagination.nextAfterCursor) {
      onEndReachedCalledDuringMomentumRef.current = true
      const { url, id } = navigation.state.params || { url: null, id: null }
      actions.loadData(true, false, url, id)
    }
  }

  const handleRefresh = () => {
    const { url, id } = navigation.state.params || { url: null, id: null }
    actions.loadData(false, true, url, id)
  }

  const avatar = (item: Item) => {
    const avatar = avatarExtractor(item)
    return {
      source: avatar && avatar.toString().startsWith('http') ? 
      { uri: avatar } : avatar ? avatar : defaultImage
    }
  }

  const handlePress = (item: Item) => {
    if (onPress) {
      onPress(item)
    } else {
      if (playlist) {
        localActions.resetAndPlayTrack(items, item.id)
      } else {
        localActions.resetAndPlayTrack([item as Track], null)
      }
    }
  }
  
  const localRenderItem: ListRenderItem<Item> = ({ item }) => {
    return (
      <ListItem
        leftAvatar={avatar(item)}
        title={titleExtractor(item)}
        titleProps={{numberOfLines: 1}}
        subtitle={subtitleExtractor(item)}
        onPress={() => handlePress(item)}
        bottomDivider
      />
    )
  }

  const { image, description } = navigation.state.params || { image: null, description: null }
  const Header =
    image && description ?
      <View style={styles.headerContainer}>
        <View style={styles.headerImageContainer}>
          <Image
            style={styles.headerImage}
            source={{uri: image}}
          />
        </View>
        { description ? <Text style={styles.headerText}>{description}</Text> : null }
      </View>
    : null
    

  return (
    <View style={styles.container}>
      <FlatList
        ListHeaderComponent={Header}
        data={items}
        renderItem={renderItem ? renderItem : localRenderItem}
        keyExtractor={keyExtractor}
        refreshing={pagination.isFetching}
        onRefresh={handleRefresh}
        onEndReachedThreshold={0.1}
        onEndReached={handleEndReached}
        onMomentumScrollBegin={() => { onEndReachedCalledDuringMomentumRef.current = false }}
      />
    </View>
  )
}

export default List
