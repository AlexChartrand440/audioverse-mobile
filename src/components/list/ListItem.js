import React from 'react'
import PropTypes from 'prop-types'
import { TouchableOpacity, View, Text, Image, Platform, StyleSheet } from 'react-native'
import Icon from 'react-native-vector-icons/Feather'

const ListItem = ({ style, avatar, icon, title, subtitle, rightElement, chevron, chevronColor, onPress }) => (
  <View>
    <PadView style={[styles.container, style]}>
      {renderIcon(icon)}
      {renderAvatar(avatar)}
      {(title || subtitle) && (
        <TouchableOpacity style={styles.contentContainer} onPress={onPress}>
          {renderNode(title, null, [styles.title])}
          {renderNode(subtitle, null, [styles.subtitle])}
        </TouchableOpacity>
      )}
      {renderNode(rightElement)}
      {chevron && <Icon name="chevron-right" size={26} color={chevronColor} onPress={onPress} />}
    </PadView>
    <View style={styles.divider} />
  </View>
)

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#FFFFFF',
    padding: Platform.OS === 'ios' ? 10 : 8
  },
  avatar: {
    width: 50,
    height: 50,
    borderRadius: 25
  },
  contentContainer: {
    flex: 1,
    justifyContent: 'center'
  },
  title: {
    fontSize: Platform.OS === 'ios' ? 17 : 16
  },
  subtitle: {
    fontSize: Platform.OS === 'ios' ? 15 : 14,
    color: '#212121'
  },
  divider: {
    // darker color if hairlineWidth is not thin enough
    backgroundColor: StyleSheet.hairlineWidth < 1 ? '#BCBBC1' : 'rgba(0, 0, 0, 0.12)',
    height: StyleSheet.hairlineWidth
  }
})

const elementOrObject = PropTypes.oneOfType([
  PropTypes.element,
  PropTypes.object
])

const elementOrString = PropTypes.oneOfType([
  PropTypes.element,
  PropTypes.string
])

ListItem.propTypes = {
  avatar: elementOrObject,
  icon: elementOrObject,
  title: elementOrString,
  subtitle: elementOrString,
  rightElement: PropTypes.element,
  chevron: PropTypes.bool,
  chevronColor: PropTypes.string,
  onPress: PropTypes.func
}

ListItem.defaultProps = {
  chevronColor: '#D1D1D6'
}

const PadView = ({ children, pad = 16, ...props }) => {
  const childrens = React.Children.toArray(children)
  const length = childrens.length
  return (
    <View {...props}>
      {React.Children.map(
        childrens,
        (child, index) =>
          child && [child, index !== length -1 && <View width={pad} />]
      )}
    </View>
  )
}

const renderAvatar = content =>
  content == null ? null : React.isValidElement(content) ? (
    content
  ) : (
    <Image {...content} source={content.source && content.source.toString().startsWith('http') ? {uri: content.source} : content.source} style={[styles.avatar, content.style]} />
  )

const renderIcon = content =>
  content == null ? null : React.isValidElement(content) ? (
    content
  ) : (
    <Icon size={24} color="rgba(0, 0, 0, 0.40)" {...content} />
  )

const renderNode = (content, props, style) =>
  content == null ? null : React.isValidElement(content) ? (
    content
  ) : (
    <Text ellipsizeMode={'tail'} numberOfLines={1} {...props} style={[style, props && props.style]}>
      {content}
    </Text>
  )

export default ListItem
