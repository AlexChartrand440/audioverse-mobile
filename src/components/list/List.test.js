import React from 'react'
import List from './List'
import ListItem from './ListItem'

import renderer from 'react-test-renderer'

describe('List component', () => {
  it('renders activity indicator when no items', () => {
    const tree = renderer.create(<List renderItem={jest.fn()} items={[]} isFetching={false} onRefresh={jest.fn()}/>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders items correctly', () => {
  const tree = renderer.create(<List renderItem={jest.fn(() => <ListItem />)} items={[{id: '1'}]} isFetching={false} onRefresh={jest.fn()}/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
