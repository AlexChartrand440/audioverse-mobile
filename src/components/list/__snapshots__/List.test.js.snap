// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component renders activity indicator when no items 1`] = `
<ActivityIndicator
  animating={true}
  color="#03A9F4"
  hidesWhenStopped={true}
  size="large"
  style={
    Object {
      "margin": 50,
    }
  }
/>
`;

exports[`List component renders items correctly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "id": "1",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.1}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[MockFunction]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[MockFunction]}
      refreshing={false}
    />
  }
  refreshing={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "flexDirection": "row",
                "padding": 10,
              },
              undefined,
            ]
          }
        />
        <View
          style={
            Object {
              "backgroundColor": "#BCBBC1",
              "height": 0.5,
            }
          }
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
